File: ranker/player.py
import math


class Player:
    """
    Represents a player or object that is to be ranked
    Contains information about rating and rating deviation
    """

    Q = math.log(10) / 400

    def __init__(self, name, *, rating=0, deviation=350):
        """
        Initializes a player with a name, rating, and deviation
        """
        self.name = name
        self.rating = rating
        self.deviation = deviation
        self.num_matches = 0

    def __repr__(self):
        return f"Player({self.name}, {self.rating}, {self.deviation}, {self.num_matches})"

    def update(self, score, opp_rating, opp_devation):
        """
        Updates the player's rating and deviation
        based on the score again an opponent
        """

        d2 = self.get_d2(opp_rating, opp_devation)
        new_rating = self.get_new_rating(score, opp_rating, opp_devation, d2)
        new_deviation = self.get_new_deviation(d2)
        self.rating = new_rating
        self.deviation = new_deviation
        self.num_matches += 1

    def get_new_rating(self, score, opp_rating, opp_deviation, d2):
        """
        Get player's new rating based on the score
        against an opponent
        """

        return self.rating + self.Q / (1 / opp_deviation**2 + 1 / d2) * (
            score - self.get_expected_score(opp_rating, opp_deviation)
        )

    def get_new_deviation(self, d2):
        """
        Get player's new deviation based on the score
        against an opponent
        """

        return math.sqrt(1 / (1 / self.deviation**2 + 1 / d2))

    def get_d2(self, opp_rating, opp_deviation):
        """
        Computes
        d^2 = 1/(q^2(g(RD')^2E(1-E)))
        where
        q = ln(10)/400
        g(RD') = 1/sqrt(1 + (3q^2(RD')^2)/pi^2)
        RD' is the rating deviation of the opponent
        E is the expected score against the opponent
        """

        expected_score = self.get_expected_score(opp_rating, opp_deviation)

        return 1 / (
            self.Q**2
            * (self.get_g(opp_deviation) ** 2)
            * expected_score
            * (1 - expected_score)
        )

    def get_expected_score(self, opp_rating, opp_deviation):
        """
        Computes the expected score of the player against the opponent
        E = 1/(1 + 10^(g(RD')(opp_rating - rating)/400))
        """

        return 1 / (
            1 + 10 ** (self.get_g(opp_deviation) * (opp_rating - self.rating) / 400)
        )

    def get_g(self, deviation):
        """
        Computes g(RD)
        g(RD) = 1/sqrt(1 + (3q^2(RD)^2)/pi^2)
        """

        return 1 / math.sqrt(1 + (3 * (self.Q**2) * (deviation**2)) / math.pi**2)

File: ranker/cpp/player.cpp
#include "player.h"

#include <cmath>

Player::Player(int id, float rating, float rd):
    id_{id}, rating_{rating}, rd_{rd}, num_comps_(0) {}

void Player::set_rating(float rating) {
    rating = rating;
}

void Player::set_rd(float rd) {
    rd = rd;
}

void Player::update(float score, float opp_rating, float opp_rd) {
    float d2 = get_d2(opp_rating, opp_rd);
    float new_rating = get_new_rating(score, opp_rating, opp_rd, d2);
    float new_rd = get_new_rd(d2);
    set_rating(new_rating);
    set_rd(new_rd);
    num_comps_++;
}

float Player::get_new_rating(float score, float opp_rating, float opp_rd, float d2) const {
    float g = get_g(opp_rd);
    return rating_ + Q / (1 / (rd_ * rd_) + 1 / d2) * g * (score - get_expected_score(opp_rating, opp_rd, g));
}

float Player::get_new_rd(float d2) const {
    return sqrt(1 / (1 / (rd_ * rd_) + 1 / d2));
}   

float Player::get_d2(float opp_rating, float opp_rd) const {
    return 1 / (Q * Q * get_g(opp_rd) * get_g(opp_rd) * get_expected_score(opp_rating, opp_rd, get_g(opp_rd)) * (1 - get_expected_score(opp_rating, opp_rd, get_g(opp_rd))));
}

float Player::get_g(float opp_rd) const {
    return 1 / sqrt(1 + 3 * Q * Q * opp_rd * opp_rd / M_PI / M_PI);
}

float Player::get_expected_score(float opp_rating, float opp_rd, float g) const {
    return 1 / (1 + pow(10, -g * (rating_ - opp_rating) / 400));
}

float Player::compute_score_var(float opp_rating, float opp_rd) const {
    float p10 = pow(10, (rating_ - opp_rating) / 400);
    float frac = (p10 / ((1 + p10) * (1 + p10)));
    float frac2 = frac * frac;
    return frac2 * (rd_ * rd_ + opp_rd * opp_rd);
}
